eq_cir_proxy_service/utils/iap.py

```diff
@@ -0,0 +1,78 @@
+"""Utility functions for handling IAP authentication and HTTP clients."""
+
+import os
+from collections.abc import AsyncIterator
+from contextlib import asynccontextmanager
+from typing import cast
+
+import google.auth.transport.requests
+import google.oauth2.id_token
+from httpx import AsyncClient
+
+
+def get_iap_token(audience: str) -> str:
+    """Fetch an ID token for the IAP-secured resource (blocking)."""
+    request = google.auth.transport.requests.Request()  # type: ignore[no-untyped-call]
+    token = google.oauth2.id_token.fetch_id_token(request, audience)  # type: ignore[no-untyped-call]
+    if token is None:
+        error_message = f"Failed to fetch IAP token for audience {audience}"
+        raise RuntimeError(error_message)
+    return cast(str, token)
+
+
+def unknown_env_error(env: str) -> str:
+    """Returns an error message for unknown environment."""
+    return f"Unknown ENV: {env}"
+
+
+@asynccontextmanager
+async def get_api_client(local_url: str, url_env: str, iap_env: str) -> AsyncIterator[AsyncClient]:
+    """Context-managed httpx.AsyncClient that switches between local and GCP/IAP.
+
+    local_url: base URL when ENV=local
+    url_env: environment variable holding the GCP base URL
+    iap_env: environment variable holding the IAP client ID
+    """
+    env = os.getenv("ENV", "local")
+
+    if env == "local":
+        client = AsyncClient(base_url=local_url)
+    elif env == "gcp":
+        base_url = os.environ[url_env]
+        audience = os.environ[iap_env]
+        client = AsyncClient(
+            base_url=base_url,
+            headers={"Authorization": f"Bearer {get_iap_token(audience)}"},
+        )
+    else:
+        raise ValueError(unknown_env_error(env))
+
+    try:
+        yield client
+    finally:
+        await client.aclose()
+
+
+# --- API-specific wrappers ---
+
+
+@asynccontextmanager
+async def get_cir_client() -> AsyncIterator[AsyncClient]:
```

Comment:
Can avoid repetition with this method and the one below by passing the vars in? Just a `get_client` method?

Reply:
In fact if you do this, you could just call `get_api_client` directly, do we need this wrapper?

Reply:
I meant this in my initial comment
```
async def get_client(local_url, url_env, iap_env) -> AsyncIterator[AsyncClient]:
    """Context-managed httpx.AsyncClient for CIR service."""
    async with get_api_client(
        local_url=local_url,
        url_env=url_env,
        iap_env=iap_env,
    ) as client:
        yield client
```
But doing this I realised there is not much point in the wrapper as far as I can see and you could just call `get_api_client` with these params? I don't think this would mean setting the values any more than times now

Reply:
I meant more than once in terms of if Proxy Service was ever updated/extended so that either CIR or Converter was called more than once. Not sure about Converter, but depending on the resolution of the CIR endpoints issue, it may be necessary to call CIR more than once. The code this way then means that the local and IAP config values don't have to be set in each CIR call, but only once in the IAP set-up. Each call then only needs to set the endpoint and any params, which are unique to the call.

Reply:
Personally I would cross that bridge when we get to it, best to try to avoid making changes with unknown/unclear scenarios in mind. It is already difficult/not possible to test this code in practice without the infrastructure bit in place so quite awkward to review in general (not your fault obviously 😅). Commented as generally I look for places where we are breaking the DRY principle and just could be made a bit simpler for right now.

Reply:
Removed repetition